### PR TITLE
 New gpg_keypair module to manage GPG keys.

### LINKED ISSUE #238: GPG module and key lookup
_From @ageis on May 09, 2019 11:17_

##### SUMMARY
Ansible could use a GPG module, for managing keyrings, importing and exporting keys, verifying and making signatures, and decrypting and encrypting data (both symmetric and non). Working with GPG keys, outside of the context of apt, is an extremely common sysadmin task which we can make it easier to automate. I think the addition of this module could be complemented by a "gpgkey" lookup which returns matching keys. I am on the fence about whether encryption, decryption, and sign+verify functions should be split into their own submodules.

I've seen no prior tickets or discussion of this, so I've taken on the task of developing it myself, and have already implemented a significant amount of the prospective functionality, and am creating this issue in order to track my progress. Please assign it to me. Hopefully this will be ready in time for 2.9. 

This is a project which could prove thorny/hairy to maintain going forward. For elucidation on why that is, I'd love to [quote](https://blog.patternsinthevoid.net/pretty-bad-protocolpeople.html) @isislovecruft: 

> "...largely due to GnuPG’s interface design which uses file descriptors, and only file descriptors, to speak a custom, potentially binary but often ascii, order dependent line protocol, whose line order, keywords, number of fields, and other details are subject to change between minor point versions of GnuPG. If that sounds like a special hell invented by some sort of unholy crossing between RMS and a rabid howler monkey: welcome to working with (or rather, more likely, around) the Terrible Idea Generator known as the GnuPG development team."

For now I've decided to rely on her version of python-gnupg which is now called pretty-bad-protoco] ([PyPi](https://pypi.org/project/pretty-bad-protocol/), [GitHub](https://github.com/isislovecruft/python-gnupg)); in my experience and the opinion of people I trust it's simply better, but that is a decision which the project/community will have to review when it's PR-time. 

I found two modules out there, but IMO neither of them are presently suitable for inclusion in core for various reasons. I want to note the prior effort by @tnt and @brandonkal anyway:
* https://github.com/tnt/ansible-gpg-import-module 
* https://github.com/brandonkal/ansible-gpg

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
- module: gpg
- lookup: gpgkey

##### ADDITIONAL INFORMATION

My work-in-progress is located here: https://git.cointel.pro/FOSS/ansible-module-gpg

_Copied from original issue: ansible/ansible#56262_

### PR DESCRIPTION
**Pull request by austinlucaslake**
##### SUMMARY

New gpg_keypair module to manage GPG keys.

Fixes #238

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

gpg_keypair

##### ADDITIONAL INFORMATION

[Unattended GPG key generation](https://www.gnupg.org/documentation/manuals/gnupg/Unattended-GPG-key-generation.html)

Example usage:

```paste below
community.crypto.gpg_keypair:
  state: present
  key_type: RSA
  key_length: 4096
  subkey_type: RSA
  subkey_length: 4096
  name: {{ name }}
  comment: {{ comment }}
  email: {{ email }}
  passphrase: {{ passphrase }}
```

### CODE REVIEW

File: plugins/modules/gpg_keypair.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    - "This module allows one to generate or delete OpenSSH private and public keys using GnuPG (gpg)."
+requirements:
+    - gpg >= 2.1
+extends_documentation_fragment:
+    - ansible.builtin.files
+    - community.crypto.attributes
+    - community.crypto.attributes.files
+attributes:
+    check_mode:
+        support: full
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:26]:
```suggestion
        support: full
    diff_mode:
        support: none
```

File: plugins/modules/gpg_keypair.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: gpg_keypair
+author: "Austin Lucas Lake (@austinlucaslake)"
+short_description: Generate or delete GPG private and public keys
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:15]:
```suggestion
short_description: Generate or delete GPG private and public keys
version_added: 2.20.0
```

File: plugins/modules/gpg_keypair.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            - Whether the private and public keys should exist or not, taking action if the state is different from what is stated.
+        type: str
+        default: present
+        choices: [ present, absent ]
+    key_type:
+        description:
+            - "Specifies the type of key to create. By default this is V(EDDSSA) which must be used with Curve25519.
+              Supported key types are V(RSA), V(DSA), V(ECDSA), V(EDDSA), and V(ECDH)."
+        type: str
+        default: EDDSA
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:39]: How about not having a default here? That way the default used by `gpg` would be used.

File: plugins/modules/gpg_keypair.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+        type: str
+        choices: ['cv25519', 'nistp256', 'nistp384', 'nistp521', 'brainpoolP256r1', 'brainpoolP384r1', 'brainpoolP512r1', 'secp256k1']
+    key_usage:
+        description:
+            - Specifies usage(s) for key.
+            - Support usages are V(encrypt), V(sign), V(auth), V(cert).
+            - V(cert) is given to all primary keys regardess, however can be used to only give V(vert) usage to a key.
+            - If not usage is specified, the valid usages for the given key type with be assigned.
+            - If O(state) is V(absent), this parameter is ignored. 
+        type: list[str]
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:66]:
```suggestion
        type: list
        elements: str
```

File: plugins/modules/gpg_keypair.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+        description:
+            - Similar to O(key_length).
+            - For ELG keys, the minimum is V(1024), the maximum is V(4096), and the default is V(3072)."
+        type: int
+    subkey_curve:
+        description:
+            - "Similar to O(key_curve)"
+        type: str
+        choices: ['cv25519', 'nistp256', 'nistp384', 'nistp521', 'brainpoolP256r1', 'brainpoolP384r1', 'brainpoolP512r1', 'secp256k1']
+    key_usage:
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:84]: Should this be `subkey_usage`?

File: plugins/modules/gpg_keypair.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        description:
+            - Similar to O(key_usage), but does not support V(cert).
+        type: list[str]
+        choices: ['encrypt', 'sign', 'auth', 'cert']
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:88]: About all the `subkey_*` options: how about having a `subkey` option which is a list of structured dicts with keys like `type`, `length`, `curve`, `key_usage`, ...? Then you can also specify multiple subkeys.

File: plugins/modules/gpg_keypair.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+    passphrase:
+        description:
+            - Passphrase used to decrypt an existing private key or encrypt a newly generated private key.
+            - If O(state) is V(absent), this parameter is ignored. 
+        type: str
+    fingerprints:
+        description:
+            - Specifies keys to match against.
+            - Provided fingerprints will take priority over user-id "V(name) (V(comment)) <V(email)>".
+            - If O(state) is V(absent), keys with the provided fingerprints will be deleted if found
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:110]:
```suggestion
            - If O(state) is V(absent), keys with the provided fingerprints will be deleted if found.
```

File: plugins/modules/gpg_keypair.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+            - If O(state) is V(absent), this parameter will be ignored.
+        type: str
+    transient_key:
+        description:
+            - Allows key generation to use a faster, but less secure random number generator.
+        type: bool
+        default: False
+    return_fingerprints:
+        description:
+            - Allows for the return of fingerprint(s) for newly created or deleted keys(s)
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:124]:
```suggestion
            - Allows for the return of fingerprint(s) for newly created or deleted keys(s).
```
But why not simply always return the fingerprints?

File: plugins/modules/gpg_keypair.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+  community.crypto.gpg_keypair:
+    state: abscent
+    fingerprints:
+      - ABC123...
+
+'''
+
+RETURN = '''
+size:
+    description: Size (in bits) of the SSH private key.
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:184]: SSH → GPG? :)

File: plugins/modules/gpg_keypair.py [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+    returned: changed or success
+    type: int
+    sample: 4096
+fingerprints:
+    description: Fingerprint(s) of newly created or deleted key(s)
+    return: changed and `return_fingerprints`==True
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:190]:
```suggestion
    return: changed and O(return_fingerprints=true)
```

File: plugins/modules/gpg_keypair.py [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+import re
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.community.crypto.plugins.module_utils.gnupg.cli import GPGError
+from ansible_collections.community.crypto.plugins.plugin_utils.gnupg import GPGError
+
+def validate_params(params: Dict[str, Union[str, int]]) -> None:
+
+    if params['override'] and params['present'] and not (params['fingerprint'] or params['name'] or params['comment'] or params['email']):
+            raise GPGError, 'To override existing keys, please provide any combination of the `fingerprint`, `name`, `comment`, and `email` parameters.'
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:207]:
```suggestion
            raise GPGError('To override existing keys, please provide any combination of the `fingerprint`, `name`, `comment`, and `email` parameters.')
```

File: plugins/modules/gpg_keypair.py [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+from typing import Dict, Union
+
+import itertools
+import re
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.community.crypto.plugins.module_utils.gnupg.cli import GPGError
+from ansible_collections.community.crypto.plugins.plugin_utils.gnupg import GPGError
+
+def validate_params(params: Dict[str, Union[str, int]]) -> None:
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:204]: Please don't use type hints. These don't work with Python 2.

File: plugins/modules/gpg_keypair.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+
+    if params['override'] and params['present'] and not (params['fingerprint'] or params['name'] or params['comment'] or params['email']):
+            raise GPGError, 'To override existing keys, please provide any combination of the `fingerprint`, `name`, `comment`, and `email` parameters.'
+    keys = ['key']
+    if params['subkey_type']:
+        keys.append('subkey')
+    for key in keys:
+        if params[f'{key}_type'] == 'EDDSA':
+            if not params[f'{key}_usage']: params[f'{key}_usage'] = ['sign', 'auth']
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:213]: `f` strings don't work with Python 2 either. You need to use `%` or `str.format()`.

File: plugins/modules/gpg_keypair.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+def run_module(params: Dict[str, Union[str, int, bool, List[str]]], check_mode: bool = False):
+    validate_params(params)
+    gpg_runner = PluginGPGRunner()
+    matching_keys = list_matching_keys(
+        params["name"],
+        params["comment"],
+        params["email"],
+        params["fingerprints"]
+    )
+    result = generate_keypair(gpg_runner, params, matching_keys, check_mode) if params['state'] == 'present' else delete_keypair(gpg_runner, matching_keys, check_mode)
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:352]: It's easier to read if you split this up:
```suggestion
    if params['state'] = present:
        result = generate_keypair(gpg_runner, params, matching_keys, check_mode)
    else:
        result = delete_keypair(gpg_runner, matching_keys, check_mode)
```

File: plugins/modules/gpg_keypair.py [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+        {f'Name-Email: {params["email"]}' if params['email'] else ''}
+        {f'Passphrase: {params["passphrase"]}' if params['passphrase'] else '%no-protection'}
+        {f'Keyserver: {params["keyserver"]}' if params['keyserver'] else ''}
+        {'%transient-key' if params['transient_key'] else ''}
+        {'%dry-run' if check_mode or not params['force'] else ''}
+        %commit
+        EOF
+        """
+
+    _, stdout, _ = gpg_runner.run_command([
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:326]:
```suggestion
    dummy, stdout, dummy2 = gpg_runner.run_command([
```

File: plugins/modules/gpg_keypair.py [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+    matching_keys: List[str],
+    check_mode: bool = False
+) -> Dict[str, Union[str, int]]:
+    if matching_keys:
+        gpg_runner.run_command([
+            f'{"dry-run" if check_mode else ""}',
+            '--batch',
+            '--yes',
+            '--delete-secret-and-public-key',
+            *matching_key
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:287]:
```suggestion
            *matching_keys
```

File: plugins/modules/gpg_keypair.py [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+    if params['name']:
+        user_id += f'{params["name"]} '
+    if params['comment']:
+        user_id += f'({params["comment"]}) '
+    if params['email']:
+        user_id += f'<{params["email"]}>'
+    if user_id:
+        user_id = f'"{user_id.strip()}"'
+
+    if user_id or fingerprints:
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:269]: The argument is called `fingerprint` (singular).

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments below.

One thing that should be discussed is idempotence. I.e. what does happen if parameters like `key_type` do not match the existing key (if `state=present`).

Generally Ansible should ensure the state, i.e. if the existing key has the wrong type, size, key usage, ..., it should be modified to the module's parameters. For objects such as private keys (or disk partitions etc.), destructive operations are generally a bad idea, so it makes sense to make that configurable, but other changes (like key usage) can be made without cycling the private key. For the private key handling modules in this collection, we usually make this behavior configurable (consider for example the `regenerate` option of `community.crypto.openssl_privatekey`).

Do you already have plans (after all this PR is marked as a draft)?

File: plugins/modules/gpg_keypair.py (hunk 5 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:87]: Yes, it should be subkey_usage. Thanks for catching that.

File: plugins/modules/gpg_keypair.py (hunk 13 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:217]: I'll need do some refactor as I used quite a few f strings haha. Refactoring will be commited later today.

**Comment by austinlucaslake** [comment]
> Thanks for your contribution! I've added some first comments below.
> 
> One thing that should be discussed is idempotence. I.e. what does happen if parameters like `key_type` do not match the existing key (if `state=present`).
> 
> Generally Ansible should ensure the state, i.e. if the existing key has the wrong type, size, key usage, ..., it should be modified to the module's parameters. For objects such as private keys (or disk partitions etc.), destructive operations are generally a bad idea, so it makes sense to make that configurable, but other changes (like key usage) can be made without cycling the private key. For the private key handling modules in this collection, we usually make this behavior configurable (consider for example the `regenerate` option of `community.crypto.openssl_privatekey`).
> 
> Do you already have plans (after all this PR is marked as a draft)?

Thanks for the comments! Regeneration as you described, would be great. However there is a problem when matching against specific key types. The `key_type` parameter can't always be extracted when listing keys. For clarification, please take a look at the example outputs below when listing keys...

This is the output for an ECDSA key using the secp256k1 curve:
```
sec   secp256k1/44A9910642E57A60 2024-05-01 [SCA]
      CF68D75C4321667AE7E815F244A9910642E57A60
uid                 [ultimate] name (comment) <name@email.com>
```
This the output for an ECDSA key using the brainpoolP512r1 curve:
```
sec   brainpoolP512r1/5A6CA22A7DA2FE9C 2024-05-01 [SCA]
      F6892C47B20B59ACB652F3B95A6CA22A7DA2FE9C
uid                 [ultimate] name (comment) <name@email.com>
```
This is the output for an ECDSA key with an ECDH subkey, both of which use brainpoolP521r1 curve:
```
sec   brainpoolP512r1/AA5C22A75FD5B23D 2024-05-01 [SCA]
      E2280C64B089AEBCD7CE4EBEAA5C22A75FD5B23D
uid                 [ultimate] name (comment) <name@email.com>
ssb   brainpoolP512r1/836D4694DAF4DE3B 2024-05-01 [E]
```

This problem really only applies for when trying to decipher between ECDSA and ECDH keys. RSA, DSA, and ELG keys all include their key_type tags in the output along with the bit length. And EDDSA is output as ed25519 as thats the only curve its compatible with. Now that I think about it, it may be possible to differentiate between ECDSA and ECDH if I factor in the key usage as well. I'll see what I can do in that regard in the next couple of days.

File: plugins/modules/gpg_keypair.py (hunk 6 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:92]: Creating GPG keys in batch mode (i.e. without user interaction) doesn't support multiple keys as is, but I could do some modifications to the code do get around that. I'll give that suggestion a try once I address some of the other things you mentioned.

File: plugins/modules/gpg_keypair.py (hunk 9 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:188]: Thought I deleted the size return value entirely. Whoops

File: plugins/modules/gpg_keypair.py (hunk 8 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:128]: I'm not 100% sure where the community would stand on this, but I had wanted to give users the option to keep the fingerprints hidden.

**Comment by felixfontein** [comment]
Did you try to look at the `--with-colons` output?

**Comment by austinlucaslake** [comment]
> Did you try to look at the `--with-colons` output?

Just took a look at that now. That should work. From what I see from the output, the OpenPGP algorithm id numbers can be parsed. I can convert those into the corresponding values for the `key_type` or `subkey_type` parameters input by users of the module.

File: plugins/modules/gpg_keypair.py (hunk 6 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:92]: So in order to implement this, I would need to use loop and repeat `gpg --batch --quick-add-key` n-times to add an arbitrary number of subkeys. The command `--quick-add-key` accepts an `algo` parameter when can be something like `rsa`, `dsa`, or `rsa4096` and the like if you also wish to specify the number of bits. For ECC, its a little different. ECC keys require you to put the desired curve into the `algo` parameter instead of a key type such as `ecdsa`, `eddsa`, `ecdh`. Here's where a problem comes up: it seems you are unable to produce ECDSA subkeys using this method. Anytime you enter one of its supported curves (i.e. `nistp256`, `nistp384`, `nistp521`, `brainpoolP256r1`, `brainpoolP384r1`, `brainpoolP512r1`, `secp256k1`). This same list of curved is also supported but ECDH, and gpg will unfortunately defaults to it anytime you entire one of aforementioned curves into the `algo` parameter.

TLDR:
A tradeoff arises that's as follows: do you want an ECDSA subkey or do you want to have an arbitrary number of subkeys?

File: plugins/modules/gpg_keypair.py (hunk 8 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py]: Ended up removing the return_fingerprints since public keys naturally show the fingerprints anyways.

File: plugins/modules/gpg_keypair.py (hunk 6 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py]: Marking this as resolved for now. I made changes to allow for multiple subkeys. The caveat mentioned above is present, so this conversion can be reopened and discussed more in need be.

**Comment by austinlucaslake** [comment]
Will be working on conditional regeneration over the next couple of days. Should be ready for review once that is fleshed out.

File: plugins/modules/gpg_keypair.py [hunk 18]
```diff
@@ -0,0 +1,477 @@
...
+            - List of subkeys with their own respective key types, lengths, curves, and usages.
+            - Similar to O(key_type), O(key_length), O(key_curve), and (key_usage).
+            - Supports ECDH and ELG keys.
+            - For both ECDH and ELG keys, the only supported usage is V(encrypt).
+            - For ECDH keys, the default curve is V(brainpoolP512r1).
+            - ECDH keys also support the V(cv25519) curve.
+            - For ELG keys, the minimum length is V(1024) bits, the maximum length is V(4096) bits, and the default length is V(3072) bits.
+        type: list
+        elements: dict
+        options:
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:77]: In `DOCUMENTATION` it's `suboptions`, not `options`:
```suggestion
        suboptions:
```

File: plugins/modules/gpg_keypair.py [hunk 19]
```diff
@@ -0,0 +1,477 @@
...
+    type: list
+    elements: str
+    sample: [ ABC123... ]
+'''
+
+import itertools
+import re
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.community.crypto.plugins.plugin_utils.gnupg.py import PluginGPGRunner
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:179]: First, the `.py` shouldn't be there, and second, this import does not work for modules, only for non-module plugins. Modules can only import things from module_utils.

For modules you'll need to create a new `GPGRunner` impementation (in `module_utils/gnupg/`, say a new file `module_runner.py`) that uses `AnsibleModule.run_command()`.

Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:179]: Quick question on your PluginGPGRunner code. I notice that its a child class that dervices from the abstract bass class of GPGRunner, which you also wrote. What's the reasoning for doing that? Why not just put the functionality PluginGPGRunner in GPGRunner.

Comment (felixfontein) [reply plugins/modules/gpg_keypair.py:179]: `PluginGPGRunner` is only acceptable for plugins, not for modules, since it doesn't (and cannot) use `AnsibleModule.run_command`.

File: plugins/modules/gpg_keypair.py [hunk 20]
```diff
@@ -0,0 +1,507 @@
...
+short_description: Generate or delete GPG private and public keys
+version_added: 2.20.0
+description:
+    - "This module allows one to generate or delete GPG private and public keys using GnuPG (gpg)."
+requirements:
+    - gpg >= 2.1
+attributes:
+    check_mode:
+        description:
+            - Can run in check_mode and return changed status prediction without modifying target.
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:24]: Either change `description` to `details`, or (better!) simply remove it. The default description is there for a reason.

File: plugins/modules/gpg_keypair.py [hunk 21]
```diff
@@ -0,0 +1,507 @@
...
+        support: full
+    diff_mode:
+        description:
+            - Will return details on what has changed (or possibly needs changing in check_mode), when in diff mode.
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:28]: Same here.

File: plugins/modules/gpg_keypair.py [hunk 22]
```diff
@@ -0,0 +1,507 @@
...
+DOCUMENTATION = '''
+---
+module: gpg_keypair
+author: "Austin Lucas Lake (@austinlucaslake)"
+short_description: Generate or delete GPG private and public keys
+version_added: 2.20.0
+description:
+    - "This module allows one to generate or delete GPG private and public keys using GnuPG (gpg)."
+requirements:
+    - gpg >= 2.1
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:20]: You need the default `attributes` docs fragment:
```suggestion
    - gpg >= 2.1
extends_documentation_fragment:
  - community.crypto.attributes
```

File: plugins/modules/gpg_keypair.py [hunk 23]
```diff
@@ -0,0 +1,507 @@
...
+            - Whether the private and public keys should exist or not, taking action if the state is different from what is stated.
+        type: str
+        default: present
+        choices: [ present, absent ]
+    key_type:
+        description:
+            - Specifies the type of key to create.
+        type: str
+        choices: [ 'RSA', 'DSA', 'ECDSA', 'EDDSA' ]
+    key_length:
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:42]: In the collection, the term `key size` is usually used instead of `key length`. GPG seems to use both terms (at least on the man page), so I would prefer to use `key_size` here (if you want you can add `key_length` as an alias).

File: plugins/modules/gpg_keypair.py [hunk 24]
```diff
@@ -0,0 +1,507 @@
...
+        description:
+            - For non-ECC keys, this specifies the number of bits in the key to create.
+            - For RSA keys, the minimum is V(1024), the maximum is V(4096), and the default is V(3072).
+            - For DSA keys, the minimum is V(768), the maximum is V(3072), and the default is V(2048).
+            - As per gpg's behavior, values below the allowed ranges will be set to the respective defaults, and values above will saturate at the maximum.
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:47]: Generally if you refer to the program name, use `GPG` (capitalized). Only if you explicitly refer to the GPG executable, use `C(gpg)` (lower-case and code-style).
```suggestion
            - As per GPG's behavior, values below the allowed ranges will be set to the respective defaults, and values above will saturate at the maximum.
```

File: plugins/modules/gpg_keypair.py [hunk 25]
```diff
@@ -0,0 +1,507 @@
...
+            - If O(key_curve=ed25519) is only supported if O(key_type=EDDSA).
+            - This is required if O(key_type=ECDSA) or O(key_type=EDDSA) and it is ignored if O(key_type=RSA) or O(key_type=DSA).
+        type: str
+        choices: [ 'nistp256', 'nistp384', 'nistp521', 'brainpoolP256r1', 'brainpoolP384r1', 'brainpoolP512r1', 'secp256k1', 'ed25519' ]
+    key_usage:
+        description:
+            - Specifies usage(s) for key.
+            - V(cert) is given to all primary keys regardess, however can be used to only give V(vert) usage to a key.
+            - If not usage is specified, all of valid usages for the given key type are assigned.
+            - O(key_usage=encr) is only supported is O(key_type=RSA).
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:62]:
```suggestion
            - O(key_usage=encr) is only supported if O(key_type=RSA).
```

File: plugins/modules/gpg_keypair.py [hunk 26]
```diff
@@ -0,0 +1,507 @@
...
+            subkey_length:
+                description:
+                    - Similar to O(key_length).
+                    - For ELG subkeys, the minimum length is V(1024) bits, the maximum length is V(4096) bits, and the default length is V(3072) bits.
+                type: int
+            subkey_curve:
+                description:
+                    - Similar to O(key_curve).
+                    - V(cv25519) is supported if subkey_type is V(ECDH).
+                    - This is required if subkey_type is V(ECDSA), V(EDDSA), or V(ECDH) and it is ignored if subkey_type is V(RSA), V(DSA), or V(ELG).
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:89]:
```suggestion
                    - This is required if subkey_type is V(ECDSA), V(EDDSA), or V(ECDH) and it is ignored if O(subkeys[].subkey_type) is V(RSA), V(DSA), or V(ELG).
```

File: plugins/modules/gpg_keypair.py [hunk 27]
```diff
@@ -0,0 +1,507 @@
...
+                type: str
+                choices: ['nistp256', 'nistp384', 'nistp521', 'brainpoolP256r1', 'brainpoolP384r1', 'brainpoolP512r1', 'secp256k1', 'ed25519', 'cv25519']
+            subkey_usage:
+                description:
+                    - Similar to O(key_usage).
+                    - V(encr) is supported if subkey_type is V(RSA), V(ECDH), or V(ELG).
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:95]:
```suggestion
                    - V(encr) is supported if O(subkeys[].subkey_type) is V(RSA), V(ECDH), or V(ELG).
```

File: plugins/modules/gpg_keypair.py [hunk 28]
```diff
@@ -0,0 +1,507 @@
...
+        default: []
+        choices: [ 'encr', 'sign', 'auth', 'cert' ]
+    subkeys:
+        description:
+            - List of subkeys with their own respective key types, lengths, curves, and usages.
+        type: list
+        elements: dict
+        default: []
+        suboptions:
+            subkey_type:
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:74]: How about using
```suggestion
            key_type:
```
here (similar below?). That this is already a subkey is clear from this being part of `subkeys`.

Using the same keys as for the primary key makes it easy to copy over the values - you only have to change the indentation (except for the `-` before the first key for the subkey).

File: plugins/modules/gpg_keypair.py [hunk 29]
```diff
@@ -0,0 +1,507 @@
...
+                choices: [ 'encr', 'sign', 'auth' ]
+    expire_date:
+        description:
+            - Sets the expire date for the key.
+            - If O(expire_date=0), the key never expires.
+            - If O(expire_date=<n>), the key expires in V(n) days.
+            - If O(expire_date=<n>w), the key expires in V(n) weeks.
+            - If O(expire_date=<n>m), the key expires in V(n) months.
+            - If O(expire_date=<n>y), the key expires in V(n) years.
+            - If left unspecified, any created GPG keys never expire.
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:109]: Is it also possible to allow providing an explicit date (say in ISO format, `yyyy-mm-dd`)? Then this option would be properly idempotent, since the expiration date would be an absolute value and not depend on today's date.

File: plugins/modules/gpg_keypair.py [hunk 30]
```diff
@@ -0,0 +1,507 @@
...
+        type: str
+    name:
+        description:
+            - Specifies a name for the key's user id.
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:113]:
```suggestion
            - Specifies a name for the key's user ID.
```
(https://docs.ansible.com/ansible/devel/dev_guide/style_guide/spelling_word_choice.html#acronyms)

File: plugins/modules/gpg_keypair.py [hunk 31]
```diff
@@ -0,0 +1,507 @@
...
+        description:
+            - Specifies a comment for the key's user id.
+        type: str
+    email:
+        description:
+            - Specifies an email for the key's user id.
+        type: str
+    passphrase:
+        description:
+            - Passphrase used to decrypt an existing private key or encr a newly generated private key.
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:125]:
```suggestion
            - Passphrase used to decrypt an existing private key or encrypt a newly generated private key.
```

File: plugins/modules/gpg_keypair.py [hunk 32]
```diff
@@ -0,0 +1,507 @@
...
+        type: str
+    fingerprints:
+        description:
+            - Specifies keys to match against.
+        type: list
+        elements: str
+        default: []
+    force:
+        description:
+            - If O(force=True), key generation is executed using the module's options, even a matching key is found.
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:135]:
```suggestion
            - If O(force=true), key generation is executed using the module's options, even a matching key is found.
```
Always use `true` and `false` for booleans outside Python code. (Documentation is not considered Python code, since it's YAML.)

File: plugins/modules/gpg_keypair.py [hunk 33]
```diff
@@ -0,0 +1,507 @@
...
+            if subkey['subkey_length']:
+                algo += str(subkey['subkey_length'])
+        elif subkey['subkey_curve']:
+            algo = subkey['subkey_curve']
+        else:
+            algo = 'default'
+
+        module.run_command(
+            [
+                '--dry-run' if check_mode else '',
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:347]: I don't think providing an empty argument is a good idea. If you want to skip it, use `None`, since `AnsibleModule.run_command()` will simply omit `None` values (as opposed to empty strings).

File: plugins/modules/gpg_keypair.py [hunk 34]
```diff
@@ -0,0 +1,507 @@
...
+            ['--list-secret-keys', '--with-colons', fingerprint],
+            executable=get_bin_path('gpg'),
+            check_rc=True
+        )
+
+        subkey_count = 0
+        uid_present = not bool(uid)
+        for line in stdout.splitlines():
+            if line[:3] == 'sec':
+                primary_key = re.search(r'.+:([0-9]+):([0-9]+):[0-9A-Z]+:[0-9]+:[0-9]+::u:+([a-z]+)[A-Z]+:+\+:+([0-9a-zA-Z]+)::0:', line)
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:399]: Why not simply split by `:`, and then access the fields?

File: plugins/modules/gpg_keypair.py [hunk 35]
```diff
@@ -0,0 +1,507 @@
...
+                    subkey_type=dict(type='str', choices=key_types),
+                    subkey_length=dict(type='int', no_log=False),
+                    subkey_curve=dict(type='str', choices=key_curves),
+                    subkey_usage=dict(type='list', elements='str', default=[], choices=key_usages[:-1])
+                ),
+                required_if=[
+                    ['subkey_type', 'ECDSA', ['subkey_curve']],
+                    ['subkey_type', 'EDDSA', ['subkey_curve']],
+                    ['subkey_type', 'ECDH', ['subkey_curve']]
+                ]
```
Comment (felixfontein) [inline plugins/modules/gpg_keypair.py:480]: Generally it's a good idea to always use trailing commas. This will reduce the size of diffs when this is ever changed.
```suggestion
                    ['subkey_type', 'ECDH', ['subkey_curve']],
                ],
```

File: plugins/modules/gpg_keypair.py (hunk 20 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:24]: I added the description cause some of the tests were throwing errors without it.

File: plugins/modules/gpg_keypair.py (hunk 21 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:28]: Same as above.

File: plugins/modules/gpg_keypair.py (hunk 34 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:399]: Much simpiler than I was thinking haha. Seems like a good change.

File: plugins/modules/gpg_keypair.py (hunk 29 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:111]: GPG supports this, but I'm still thinking of a way to verify date formats nicely. For example Python 3.7 and onwards has `datetime.fromisoformat()` which could be wrapped in a try/except block in case of invalid date formats. Is there such a function for Python 2.6/2.7 for people running this on older versions?

File: plugins/modules/gpg_keypair.py (hunk 28 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:76]: Sounds good. Did that originally cause of name collision in documentation when using O(), but one of your other eliminates that issue.

File: plugins/modules/gpg_keypair.py (hunk 29 above)
Comment (austinlucaslake) [reply plugins/modules/gpg_keypair.py:111]: Marking as resolved. Seems like the `dateutil` package will satisfy this functionality.

Comment (felixfontein) [reply plugins/modules/gpg_keypair.py:114]: Please do not use dateutil, it's nt part of the Python standard library. You can use `datetime.datetime.strptime()` with an appropriate format string (`%Y-%m-%d`).
